src/slice-assign.c

```diff
@@ -37,13 +38,28 @@ r_obj* vec_assign_opts(r_obj* x,
   obj_check_vector(x, assign_opts.x_arg, assign_opts.call);
   obj_check_vector(value, assign_opts.value_arg, assign_opts.call);
 
-  const struct location_opts location_opts = new_location_opts_assign();
-  index = KEEP(vec_as_location_opts(
-    index,
-    vec_size(x),
-    KEEP(vec_names(x)),
-    &location_opts
-  ));
+  const r_ssize x_size = vec_size(x);
+
+  switch (assign_opts.index_style) {
+  case VCTRS_INDEX_STYLE_location: {
+    r_obj* x_names = KEEP(vec_names(x));
+    const struct location_opts location_opts = new_location_opts_assign();
+    index = vec_as_location_opts(
+      index,
+      x_size,
+      x_names,
+      &location_opts
+    );
+    FREE(1);
+    break;
+  }
+  case VCTRS_INDEX_STYLE_condition: {
+    vec_check_condition_index(index, x_size, vec_args.i, assign_opts.call);
+    break;
+  }
```

Comment:
It was pretty easy to share the `vec_assign()` code path and just branch in a few key places